include/sys/notify.h

```diff
@@ -0,0 +1,361 @@
+
+/*
+ * Mask isolating the bits of sys_notify::flags that are available
+ * for extension.
+ */
+#define SYS_NOTIFY_EXTENSION_MASK (~BIT_MASK(SYS_NOTIFY_EXTENSION_POS))
+
+/**
+ * @brief Generic signature used to notify of result completion by
+ * callback.
+ *
+ * Functions with this role may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-ok and not sleep.
+ *
+ * Parameters that should generally be passed to such functions include:
+ *
+ * * a pointer to a specific client request structure, i.e. the one
+ *   that contains the sys_notify structure.
+ * * the result of the operation, either as passed to
+ *   sys_notify_finalize() or extracted afterwards using
+ *   sys_notify_fetch_result().  Expected values are
+ *   service-specific, but the value shall be non-negative if the
+ *   operation succeeded, and negative if the operation failed.
+ */
+typedef void (*sys_notify_generic_callback)();
+
+/**
+ * @brief State associated with notification for an asynchronous
+ * operation.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. sys_notify_init_signal()) to
+ * configure them.  Generally the structure is a member of a
+ * service-specific client structure, such as onoff_client.
+ *
+ * Control of the containing object transfers to the service provider
+ * when a pointer to the object is passed to a service function that
+ * is documented to take control of the object, such as
+ * onoff_service_request().  While the service provider controls the
+ * object the client must not change any object fields.  Control
+ * reverts to the client:
+ * * if the call to the service API returns an error;
+ * * when operation completion is posted.  This may occur before the
+ *   call to the service API returns.
+ *
+ * Operation completion is technically posted when the flags field is
+ * updated so that sys_notify_fetch_result() returns success.  This
+ * will happen before the signal is posted or callback is invoked.
+ * Note that although the manager will no longer reference the
+ * sys_notify object past this point, the containing object may have
+ * state that will be referenced within the callback.  Where callbacks
+ * are used control of the containing object does not revert to the
+ * client until the callback has been invoked.  (Re-use within the
+ * callback is explicitly permitted.)
+ *
+ * After control has reverted to the client the notify object must be
+ * reinitialized for the next operation.
+ *
+ * The content of this structure is not public API to clients: all
+ * configuration and inspection should be done with functions like
+ * sys_notify_init_callback() and sys_notify_fetch_result().
+ * However, services that use this structure may access certain
+ * fields directly.
+ */
+struct sys_notify {
+	union method {
+		/* Pointer to signal used to notify client.
```

Comment:
Documentation comment here `/**`

Reply:
It's not public API (it's manipulated through accessor functions).  And documentation for struct members isn't generated so it wouldn't help.